### ISSUE TITLE
setContext not being called for query

### ISSUE THREAD

**Issue by Johnhhorton** [apollographql/apollo-client#6623]
**Intended outcome:**
`setContext`'s setter should be called on every request allowing me to dynamically set headers after an SPA login.

**Actual outcome:**
Console logs that describe what events are occurring.

Notice that setContext is called twice during the signin mutation. Note that is happens whether or not the second query is run.

Second to last log SHOULD be `running setContext`
![image](https://user-images.githubusercontent.com/12238902/87725738-d9ff7300-c78b-11ea-9a01-05bde86af6df.png)

**How to reproduce the issue:**

```typescript
import { ApolloClient, ApolloLink, HttpLink } from "@apollo/client";
import { setContext } from "@apollo/client/link/context";
import { loader } from "graphql.macro";
import { cache } from "./cache";

const httpLink = new HttpLink({
  uri: "http://localhost:3001/graphql",
});

const authLink = setContext((_, { headers, ...context }) => {
  console.log("running setContext");
  return {
    headers: {
      ...headers,
      authorization: localStorage.getItem("token")
        ? `Bearer ${localStorage.getItem("token")}`
        : "nothing here", // set to verify it was setting in headers
    },
    ...context,
  };
});

export const apolloClient = new ApolloClient({
  link: ApolloLink.from([authLink, httpLink]),
  cache,
  typeDefs: loader("./schema.graphql"),
});

```

Code that is being called:
```typescript
export const getMe = async () => {
  console.log("before running new query after token set");
  const meResult = await apolloClient.query<MeQuery>({
    query: MeDocument,
  });
  const me = meResult.data?.me;

  appState({ ...appState(), me });
};

export const signIn = async (input: { email: string; password: string }) => {
  try {
    console.log("before signin mutation");
    const result = await apolloClient.mutate<SignInMutation>({
      mutation: SignInDocument,
      variables: {
        input,
      },
    });
    console.log("succesfully passed signin mutation");

    const token = result.data?.signIn.access_token!;
    localStorage.setItem("token", token);

  // settings token in state, but not using it in the setContext.
    appState({ ...appState(), token });
    console.log("Token has been set and verified to exist");
    await getMe();
  } catch (e) {
    console.log("new query has thrown. authorization was not set in header.");
    // console.log("caught signin error", e);
    throw e;
  }

  return true;
};
```
headers for the Me query
![image](https://user-images.githubusercontent.com/12238902/87726992-17fd9680-c78e-11ea-8c0d-19a24b676bee.png)

**Versions**
```
 System:
    OS: macOS 10.15.5
  Binaries:
    Node: 12.16.3 - ~/.nvm/versions/node/v12.16.3/bin/node
    Yarn: 1.22.4 - /usr/local/bin/yarn
    npm: 6.14.4 - ~/.nvm/versions/node/v12.16.3/bin/npm
  Browsers:
    Chrome: 83.0.4103.116
    Edge: 83.0.478.64
    Firefox: 77.0.1
    Safari: 13.1.1
  npmPackages:
    @apollo/client: ^3.0.1 => 3.0.1 
```

**Comment by benjamn** [comment]
I believe this PR will address the underlying issue here: https://github.com/apollographql/apollo-client/pull/6656

The underlying issue: `@apollo/client/link/context` currently has its own copy of the `ApolloLink` class, which is a duplicate of the one known to `@apollo/client`.

**Comment by benjamn** [comment]
Heads up: we're not quite ready to publish `@apollo/client@3.1.0` yet, but I've published an `@apollo/client@3.1.0-pre.0` release that I believe will resolve this issue, if you want to try that in the meantime. 🙏

**Comment by benjamn** [comment]
@Johnhhorton This should now be fixed in `@apollo/client@3.1.0`, assuming #6656 was the solution to your problem. I'll leave this issue open so you can confirm. Thanks!

**Comment by Johnhhorton** [comment]
@benjamn It's working! No longer have to do a `window.location` change.

Thanks so much 👍  Love the tooling ❤️
